README.md
winrm-s
=======

`winrm-s` extends the functionality of the
[WinRM gem](http://rubygems.org/gems/winrm) to support the Microsoft
[Negotiate protocol](http://msdn.microsoft.com/en-us/library/windows/desktop/aa378748(v=vs.85).aspx)
when authenticating to a remote [WinRM](http://msdn.microsoft.com/en-us/library/aa384426(v=vs.85).aspx) endpoint from a Windows system.

This extended functionality is **only** supported when running on Microsoft
Windows. This gem can still be used on other operating systems just like the
`WinRM` gem, but the extended capabilities will not be available.

Installation
------------

To install it, run:

    gem install winrm-s

Usage
-----

`winrm-s` provides the same interface as the `winrm` gem -- see `winrm`
[documentation](https://github.com/WinRb/WinRM/blob/master/README.md) for `winrm-s` usage.

* To use it, simply require `winrm` or `winrm-s`, depending on whether your code
is running on Windows. The extended negotiate protocol is only available if
you include `winrm-s`, which will only work on Windows.
* When you use WinRM::WinRMWebService.new, be sure to specify the
  `:sspinegotiate` parameter, along with a user name in the form `domain_name\user_name`
  for the user name in order to make use of negotiate protocol. If the user
  account is local to the remote system, you can use `.` for the domain. The
  example further on demonstrates the negotiate use case.
* All other use cases enabled by the `winrm` gem are also supported.

Example
-------
Note the argument value of `:sspinegotiate` for transport option, and the
explicit specification of a domain name, in this case `.`, in the user name:
```ruby
if RUBY_PLATFORM =~ /mswin|mingw32|windows/
  require 'winrm-s' # only works on Windows, otherwise use require 'winrm'
  endpoint = http://mywinrmhost:5985/wsman
  winrm = WinRM::WinRMWebService.new(endpoint, :sspinegotiate, :user => ".\administrator", :pass => "adminpasswd")
  winrm.cmd('ipconfig /all') do |stdout, stderr|
    STDOUT.print stdout
    STDERR.print stderr
  end
end
```

License
-------

Copyright:: Copyright (c) 2014 Chef Software, Inc.
License:: Apache License, Version 2.0

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.



CHANGELOG.md
# winrm-s  Change Log
Release: 0.3.7.dev
--------------
* Fixes output parsing when output spans multiple responses

Release: 0.3.6
--------------
* Fixes syntax error in winrm initializer

Release: 0.3.5
--------------
* Fixes incompatibility with winrm 1.6 where a consuming gem sets the transport option to `:negotiate` and results in an error since this gem has no awareness of that setting

Release: 0.3.4
--------------
* [Issue #32](https://github.com/chef/winrm-s/issues/32): Fixes incompatibility with winrm 1.5

Release: 0.3.3
--------------
* Do not emit warnings for either httpclient versions 2.6 or 2.7

Release: 0.3.2
--------------
* Strip UTF-8 BOM from response added by win2k8R2 since newer WinRM versions now use UTF-8
* Decode all authorized payloads, not just successful ones so an unsuccesful command error can be communicated

Release: 0.3.1
--------------
* [Issue #27](https://github.com/chef/winrm-s/pull/27): Fixes problem with AllowUnencrypted being true and using negotiate

Release: 0.3.0
--------------
* [Issue #21](https://github.com/chef/winrm-s/issues/21): Bumps winrm dependency from 1.2 to 1.3.
* [Issue #25](https://github.com/opscode/winrm-s/pull/25): [Fix for knife-windows issue 218](https://github.com/opscode/knife-windows/issues/218)

Release: 0.2.4
--------------
* DOC change only: correct omission from CHANGELOG items in 0.2.3 release

Release: 0.2.3
--------------
* [Issue #20](https://github.com/chef/winrm-s/pull/20): Allow support for httpclient 2.6

Release: 0.2.2
--------------
* [Issue #16](https://github.com/opscode/winrm-s/pull/16): [fix for knife-windows issue 124](https://github.com/opscode/knife-windows/issues/124)
* [Issue #15](https://github.com/opscode/winrm-s/pull/15): drop Ruby 1.9.2 builds from Travis

Release: 0.2.0
--------------
* Fix invalid use of frozen string ENV block is used for user and domain
* Update to WinRM 1.2.0 gem dependency

Release: 0.1.0
--------------
* Initial implementation


spec/spec_helper.rb
$:.unshift File.expand_path('../../lib', __FILE__)

require 'nokogiri'

def windows?
  !!(RUBY_PLATFORM =~ /mswin|mingw|windows/)
end

def unix?
  !windows?
end

def windows_and_func_spec_enabled?
  windows? and ENV["enable_winrms_func_spec"]
end

RSpec.configure do |config|
  config.filter_run_excluding :windows_only => true unless windows?
  config.filter_run_excluding :windows_and_func_spec_enabled => true unless windows_and_func_spec_enabled?
  config.filter_run_excluding :unix_only => true unless unix?
end

spec/unit/platform_spec.rb
require 'spec_helper'


describe "Test if patch is applied on Platform" do

  describe 'Windows is succesfully patched for transport sspinegotiate', :windows_only do
    before do
      require 'winrm-s'
      @winrm = WinRM::WinRMWebService.new("http://mywinrmhost:5985/wsman", :sspinegotiate, :user => "test_winrm_user", :pass => "test_winrm_pass")
    end

    it 'HTTP::HttpSSPINegotiate class should exist' do
      expect{WinRM::HTTP::HttpSSPINegotiate}.not_to raise_exception
    end

    it 'should patch httpclient to contain encrypt/decrypt methods' do
      expect(HTTPClient::SSPINegotiateAuth.new).to respond_to(:encrypt_payload)
      expect(HTTPClient::SSPINegotiateAuth.new).to respond_to(:decrypt_payload)
    end

  end

  describe 'Winrm cannot be patched for unix', :unix_only do
    it 'require winrm-s should raise exception' do
      expect{require 'winrm-s'}.to raise_exception RuntimeError
    end
  end
end


spec/functional/sspi_nego_encrypt_decrypt_spec.rb
require 'spec_helper'

def assert_prerequisites
  if windows?
    require 'winrm-s'
    ["test_winrm_endpoint", "test_winrm_user", "test_winrm_pass"].each do |env_var|
      raise "\n\nError: Please set the environment variable: #{env_var}\n\n" if ENV[env_var].nil?
    end
  end
end


describe "Test sspinegotiate with encrypt/decrypt via WinRM" do
  before(:all) do
    assert_prerequisites
    # Remember the user setting.
    winrm_cfg = %x{winrm get winrm/config/service -format:xml}
    doc = Nokogiri::XML(winrm_cfg)
    @original_allowunencrypted_val = doc.at_css("cfg|Service cfg|AllowUnencrypted").text.to_s
  end

  before do
    %x{winrm set winrm/config/service @{AllowUnencrypted="false"}}
    @winrm = WinRM::WinRMWebService.new(ENV["test_winrm_endpoint"], :sspinegotiate, :user => ENV["test_winrm_user"].dup, :pass => ENV["test_winrm_pass"].dup)
  end

  after(:all) do
    %x{winrm set winrm/config/service @{AllowUnencrypted="#{@original_allowunencrypted_val}"}}
  end

  it 'should run a CMD command string' do
    output = @winrm.run_cmd('ipconfig /all')
    expect(output[:exitcode]).to be_zero
    expect(output[:data]).not_to be_empty
  end

  it 'should run a CMD command with proper arguments' do
    output = @winrm.run_cmd('ipconfig', %w{/all})
    expect(output[:exitcode]).to be_zero
    expect(output[:data]).not_to be_empty
  end

  it 'should run a CMD command with block' do
    outvar = ''
    @winrm.run_cmd('ipconfig', %w{/all}) do |stdout, stderr|
      outvar << stdout
    end
    expect(outvar).to match(/Windows IP Configuration/)
  end
end


lib/winrm-s.rb
#
# Author:: Kaustubh Deorukhkar (<kaustubh@clogeny.com>)>
# Copyright:: Copyright (c) 2014 Chef Software, Inc.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'winrm'

def windows?
  !!(RUBY_PLATFORM =~ /mswin|mingw|windows/)
end

# Patch only on windows
if windows?
  require 'winrm/winrm_service_patch'
else
  raise "ERROR: winrm-s extensions to the winrm gem for the negotiate protocol are only supported on Windows. Require 'winrm' and not 'winrm-s' on non-Windows platforms."
end


lib/winrm/win32/sspi.rb

require 'winrm/helpers/assert_patch'

def validate_patch
  # The code below patches the Win32::SSPI module from ruby core to add support
  # for encrypt/decrypt as described below.
  # Add few restrictions to make sure the patched methods are still
  # available, but still give a way to consciously use later versions
  PatchAssertions.assert_arity_of_patched_method(Win32::SSPI::NegotiateAuth, "initialize", -1)
  PatchAssertions.assert_arity_of_patched_method(Win32::SSPI::NegotiateAuth, "complete_authentication", 1)
  PatchAssertions.assert_arity_of_patched_method(Win32::SSPI::NegotiateAuth, "get_credentials", 0)
end

# Perform the patch validations
validate_patch

# Overrides and enhances the ruby core win32 sspi module to add support to
# encrypt/decrypt data to be sent over channel, example using SSP Negotiate auth

module Win32
  module SSPI
    # QueryContextAttributes attributes flags
    SECPKG_ATTR_SIZES = 0x00000000

    module API
      QueryContextAttributes = Win32API.new("secur32", "QueryContextAttributes", 'pLp', 'L')
      EncryptMessage = Win32API.new("secur32", "EncryptMessage", 'pLpL', 'L')
      DecryptMessage = Win32API.new("secur32", "DecryptMessage", 'ppLp', 'L')
    end

    class SecPkgContext_Sizes
      attr_accessor :cbMaxToken, :cbMaxSignature, :cbBlockSize, :cbSecurityTrailer

      def initialize
        @cbMaxToken = @cbMaxSignature = @cbBlockSize = @cbSecurityTrailer = 0
      end

      def cbMaxToken
        @cbMaxToken = @struct.unpack("LLLL")[0] if @struct
      end

      def cbMaxSignature
        @cbMaxSignature = @struct.unpack("LLLL")[1] if @struct
      end

      def cbBlockSize
        @cbBlockSize = @struct.unpack("LLLL")[2] if @struct
      end

      def cbSecurityTrailer
        @cbSecurityTrailer = @struct.unpack("LLLL")[3] if @struct
      end

      def to_p
        @struct ||= [@cbMaxToken, @cbMaxSignature, @cbBlockSize, @cbSecurityTrailer].pack("LLLL")
      end
    end

    #  SecurityBuffer for data to be encrypted
    class EncryptedSecurityBuffer

      SECBUFFER_DATA = 0x1    # Security buffer data
      SECBUFFER_TOKEN = 0x2   # Security token
      SECBUFFER_VERSION = 0

      def initialize(data_buffer, sizes)
        @original_msg_len = data_buffer.length
        @cbSecurityTrailer = sizes.cbSecurityTrailer
        @data_buffer = data_buffer
        @security_trailer = "\0" * sizes.cbSecurityTrailer
      end

      def to_p
        # Assumption is that when to_p is called we are going to get a packed structure. Therefore,
        # set @unpacked back to nil so we know to unpack when accessors are next accessed.
        @unpacked = nil
        # Assignment of inner structure to variable is very important here. Without it,
        # will not be able to unpack changes to the structure. Alternative, nested unpacks,
        # does not work (i.e. @struct.unpack("LLP12")[2].unpack("LLP12") results in "no associated pointer")
        @sec_buffer = [@original_msg_len, SECBUFFER_DATA, @data_buffer, @cbSecurityTrailer, SECBUFFER_TOKEN, @security_trailer].pack("LLPLLP")
        @struct ||= [SECBUFFER_VERSION, 2, @sec_buffer].pack("LLP")
      end

      def buffer
        unpack
        @buffer
      end

    private

      # Unpacks the SecurityBufferDesc structure into member variables. We
      # only want to do this once per struct, so the struct is deleted
      # after unpacking.
      def unpack
        if ! @unpacked && @sec_buffer && @struct
          dataBufferSize, dType, dataBuffer, tokenBufferSize, tType, tokenBuffer = @sec_buffer.unpack("LLPLLP")
          dataBufferSize, dType, dataBuffer, tokenBufferSize, tType, tokenBuffer = @sec_buffer.unpack("LLP#{dataBufferSize}LLP#{tokenBufferSize}")
          # Form the buffer stream as required by server
          @buffer = [tokenBufferSize].pack("L")
          @buffer << tokenBuffer << dataBuffer
          @struct = nil
          @sec_buffer = nil
          @unpacked = true
        end
      end
    end

    class DecryptedSecurityBuffer

      SECBUFFER_DATA = 0x1    # Security buffer data
      SECBUFFER_TOKEN = 0x2   # Security token
      SECBUFFER_VERSION = 0

      def initialize(buffer)
        # unpack to extract the msg and token
        token_size, token_buffer, enc_buffer = buffer.unpack("L")
        @original_msg_len = buffer.length - token_size - 4
        @cbSecurityTrailer, @security_trailer, @data_buffer = buffer.unpack("La#{token_size}a#{@original_msg_len}")
      end

      def to_p
        # Assumption is that when to_p is called we are going to get a packed structure. Therefore,
        # set @unpacked back to nil so we know to unpack when accessors are next accessed.
        @unpacked = nil
        # Assignment of inner structure to variable is very important here. Without it,
        # will not be able to unpack changes to the structure. Alternative, nested unpacks,
        # does not work (i.e. @struct.unpack("LLP12")[2].unpack("LLP12") results in "no associated pointer")
        @sec_buffer = [@original_msg_len, SECBUFFER_DATA, @data_buffer, @cbSecurityTrailer, SECBUFFER_TOKEN, @security_trailer].pack("LLPLLP")
        @struct ||= [SECBUFFER_VERSION, 2, @sec_buffer].pack("LLP")
      end

      def buffer
        unpack
        @buffer
      end

    private

      # Unpacks the SecurityBufferDesc structure into member variables. We
      # only want to do this once per struct, so the struct is deleted
      # after unpacking.
      def unpack
        if ! @unpacked && @sec_buffer && @struct
          dataBufferSize, dType, dataBuffer, tokenBufferSize, tType, tokenBuffer = @sec_buffer.unpack("LLPLLP")
          dataBufferSize, dType, dataBuffer, tokenBufferSize, tType, tokenBuffer = @sec_buffer.unpack("LLP#{dataBufferSize}LLP#{tokenBufferSize}")
          @buffer = dataBuffer
          @struct = nil
          @sec_buffer = nil
          @unpacked = true
        end
      end
    end

    class NegotiateAuth
      # Override to remember password
      # Creates a new instance ready for authentication as the given user in the given domain.
      # Defaults to current user and domain as defined by ENV["USERDOMAIN"] and ENV["USERNAME"] if
      # no arguments are supplied.
      def initialize(user = nil, domain = nil, password = nil)
        if user.nil? && domain.nil? && ENV["USERNAME"].nil? && ENV["USERDOMAIN"].nil?
          raise "A username or domain must be supplied since they cannot be retrieved from the environment"
        end
        @user = user || ENV["USERNAME"].dup
        @domain = domain || ENV["USERDOMAIN"].dup
        @password = password
      end

      # Takes a token and gets the next token in the Negotiate authentication chain. Token can be Base64 encoded or not.
      # The token can include the "Negotiate" header and it will be stripped.
      # Does not indicate if SEC_I_CONTINUE or SEC_E_OK was returned.
      # Token returned is Base64 encoded w/ all new lines removed.
      def complete_authentication(token)
        raise "This object is no longer usable because its resources have been freed." if @cleaned_up

        # Nil token OK, just set it to empty string
        token = "" if token.nil?

        if token.include? "Negotiate"
          # If the Negotiate prefix is passed in, assume we are seeing "Negotiate <token>" and get the token.
          token = token.split(" ").last
        end

        if token.include? B64_TOKEN_PREFIX
          # indicates base64 encoded token
          token = token.strip.unpack("m")[0]
        end

        outputBuffer = SecurityBuffer.new
        result = SSPIResult.new(API::InitializeSecurityContext.call(@credentials.to_p, @context.to_p, nil,
          REQUEST_FLAGS, 0, SECURITY_NETWORK_DREP, SecurityBuffer.new(token).to_p, 0,
          @context.to_p,
          outputBuffer.to_p, @contextAttributes, TimeStamp.new.to_p))

        if result.ok? then
          @auth_successful = true
          return encode_token(outputBuffer.token)
        else
          raise "Error: #{result.to_s}"
        end
      ensure
        # need to make sure we don't clean up if we've already cleaned up.
        # XXX - clean up later since encrypt/decrypt needs this
        # clean_up unless @cleaned_up
      end

      def encrypt_payload(body)
        if @auth_successful
          # Approach - http://msdn.microsoft.com/en-us/magazine/cc301890.aspx
          sizes = SecPkgContext_Sizes.new
          result = SSPIResult.new(API::QueryContextAttributes.call(@context.to_p, SECPKG_ATTR_SIZES, sizes.to_p))

          outputBuffer = EncryptedSecurityBuffer.new(body, sizes)
          result = SSPIResult.new(API::EncryptMessage.call(@context.to_p, 0, outputBuffer.to_p, 0 ))
          encrypted_msg = outputBuffer.buffer

          body = <<-EOF
--Encrypted Boundary\r
Content-Type: application/HTTP-SPNEGO-session-encrypted\r
OriginalContent: type=application/soap+xml;charset=UTF-8;Length=#{encrypted_msg.length - sizes.cbSecurityTrailer - 4}\r
--Encrypted Boundary\r
Content-Type: application/octet-stream\r
#{encrypted_msg}--Encrypted Boundary\r
EOF
        end
        body
      end

      def decrypt_payload(body)
        if @auth_successful

          matched_data = /--Encrypted Boundary\s+Content-Type:\s+application\/HTTP-SPNEGO-session-encrypted\s+OriginalContent:\s+type=\S+Length=(\d+)\s+--Encrypted Boundary\s+Content-Type:\s+application\/octet-stream\s+([\S\s]+)--Encrypted Boundary/.match(body)
          raise "Error: Unencrypted communication not supported. Please check winrm configuration winrm/config/service AllowUnencrypted flag." if matched_data.nil?

          encrypted_msg = matched_data[2]

          outputBuffer = DecryptedSecurityBuffer.new(encrypted_msg)
          result = SSPIResult.new(API::DecryptMessage.call(@context.to_p, outputBuffer.to_p, 0, 0 ))
          body = outputBuffer.buffer
        end
        body
      end

      private
      # ** Override to add password support
      # Gets credentials based on user, domain or both. If both are nil, an error occurs
      def get_credentials
        @credentials = CredHandle.new
        ts = TimeStamp.new
        @identity = Identity.new @user, @domain, @password
        result = SSPIResult.new(API::AcquireCredentialsHandle.call(nil, "Negotiate", SECPKG_CRED_OUTBOUND, nil, @identity.to_p,
          nil, nil, @credentials.to_p, ts.to_p))
        raise "Error acquire credentials: #{result}" unless result.ok?
      end
    end
  end
end


lib/winrm/winrm_service_patch.rb

require 'winrm/http/transport_patch'

module WinRM
  class WinRMWebService

    # Override winrm to support sspinegotiate option.

    # @param [String,URI] endpoint the WinRM webservice endpoint
    # @param [Symbol] transport either :kerberos(default)/:ssl/:plaintext
    # @param [Hash] opts Misc opts for the various transports.
    #   @see WinRM::HTTP::HttpTransport
    #   @see WinRM::HTTP::HttpGSSAPI
    #   @see WinRM::HTTP::HttpSSL
    def initialize(endpoint, transport = :kerberos, opts = {})
      @endpoint = endpoint
      @timeout = DEFAULT_TIMEOUT
      @max_env_sz = DEFAULT_MAX_ENV_SIZE 
      @locale = DEFAULT_LOCALE
      @logger = Logging.logger[self]
      case transport
      when :kerberos
        require 'gssapi'
        # TODO: check fo keys and throw error if missing
        @xfer = HTTP::HttpGSSAPI.new(endpoint, opts[:realm], opts[:service], opts[:keytab], opts)
      when :plaintext
        @xfer = HTTP::HttpPlaintext.new(endpoint, opts[:user], opts[:pass], opts)
      when :sspinegotiate
        @xfer = HTTP::HttpSSPINegotiate.new(endpoint, opts[:user], opts[:pass], opts)
      when :ssl
        @xfer = HTTP::HttpSSL.new(endpoint, opts[:user], opts[:pass], opts[:ca_trust_path], opts)
      when :negotiate
        @xfer = HTTP::HttpNegotiate.new(endpoint, opts[:user], opts[:pass], opts)
      end
    end

    # Get the builder obj for output request
    # @param [String] shell_id The shell id on the remote machine.  See #open_shell
    # @param [String] command_id The command id on the remote machine.  See #run_command
    # @return [Builder::XmlMarkup] Returns a Builder::XmlMarkup for request message 
    def get_builder_obj(shell_id, command_id, &block)
      body = { "#{NS_WIN_SHELL}:DesiredStream" => 'stdout stderr',
        :attributes! => {"#{NS_WIN_SHELL}:DesiredStream" => {'CommandId' => command_id}}}
      builder = Builder::XmlMarkup.new
      builder.instruct!(:xml, :encoding => 'UTF-8')
      builder.tag! :env, :Envelope, namespaces do |env|
        env.tag!(:env, :Header) { |h| h << Gyoku.xml(merge_headers(header,resource_uri_cmd,action_receive,selector_shell_id(shell_id))) }
        env.tag!(:env, :Body) do |env_body|
          env_body.tag!("#{NS_WIN_SHELL}:Receive") { |cl| cl << Gyoku.xml(body) }
        end
      end
      builder
    end

    # Get the Output of the given shell and command
    # @param [String] shell_id The shell id on the remote machine.  See #open_shell
    # @param [String] command_id The command id on the remote machine.  See #run_command
    # @return [Hash] Returns a Hash with a key :exitcode and :data.  Data is an Array of Hashes where the cooresponding key
    #   is either :stdout or :stderr.  The reason it is in an Array so so we can get the output in the order it ocurrs on
    #   the console.
    def get_command_output(shell_id, command_id, &block)
      done_elems = []
      output = Output.new

      while done_elems.empty?
        resp_doc = nil
        builder = get_builder_obj(shell_id, command_id, &block)
        request_msg = builder.target!
        resp_doc = send_get_output_message(request_msg)

        REXML::XPath.match(resp_doc, "//#{NS_WIN_SHELL}:Stream").each do |n|
          next if n.text.nil? || n.text.empty?
          stream = { n.attributes['Name'].to_sym => Base64.decode64(n.text).force_encoding('utf-8').sub("\xEF\xBB\xBF", "") }
          output[:data] << stream
          yield stream[:stdout], stream[:stderr] if block_given?
        end

        # We may need to get additional output if the stream has not finished.
        # The CommandState will change from Running to Done like so:
        # @example
        #   from...
        #   <rsp:CommandState CommandId="..." State="http://schemas.microsoft.com/wbem/wsman/1/windows/shell/CommandState/Running"/>
        #   to...
        #   <rsp:CommandState CommandId="..." State="http://schemas.microsoft.com/wbem/wsman/1/windows/shell/CommandState/Done">
        #     <rsp:ExitCode>0</rsp:ExitCode>
        #   </rsp:CommandState>
        done_elems = REXML::XPath.match(resp_doc, "//*[@State='http://schemas.microsoft.com/wbem/wsman/1/windows/shell/CommandState/Done']")
      end
      output[:exitcode] = REXML::XPath.first(resp_doc, "//#{NS_WIN_SHELL}:ExitCode").text.to_i
      output
    end
  end
end

lib/winrm/helpers/assert_patch.rb

module PatchAssertions

  def self.assert_major_version(library_name, expected_major, override_var_name)
    supported_major_ver = ENV[override_var_name] || expected_major
    loaded_ver = Gem.loaded_specs[library_name].version
    loaded_major_ver = loaded_ver.to_s.match(/(^\d+\.\d+)\.(.*)$/)[1]

    version_match = false
    Array(supported_major_ver).each do |version|
      if loaded_major_ver.to_s == version.to_s
        version_match = true
      end
    end
    unless version_match
      puts "WARNING: Unsupported version of #{library_name}. The supported major version of library is #{library_name} version #{expected_major}. This code path monkey patches few methods in #{library_name} to support additional features. This could possibly work, but it is advised to extensively test your version. If you are aware of the impact of using #{loaded_ver}, this warning can be disabled by setting #{override_var_name} to the major version #{loaded_major_ver}. "
    end
  end

  def self.assert_arity_of_patched_method(klass, method_name, expected_arity)
    if klass.instance_method(method_name).arity != expected_arity
      puts "WARNING: Cannot patch method #{klass}::#{method_name} since your latest gem seems to have different method definition that cannot be safely patched. This could possibly work, but it is advised to extensively test your version. Please use the supported version of patched gem."
    end
  end
end

lib/winrm/http/transport_patch.rb
#
# Author:: Kaustubh Deorukhkar (<kaustubh@clogeny.com>)
# Copyright:: Copyright (c) 2014 Chef Software, Inc.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

module WinRM
  module HTTP

    class HttpSSPINegotiate < HttpTransport
      def initialize(endpoint, user, pass, opts)
        # Override the relevant functionality in httpclient to make sspi work.
        require 'winrm/http/auth'
        super(endpoint)
        @httpcli.set_auth(nil, user, pass)
        # Remove non-sspi auths
        auths = @httpcli.www_auth.instance_variable_get('@authenticator')
        auths.delete_if {|i| not i.is_a?(HTTPClient::SSPINegotiateAuth)}
      end
    end

  end
end # WinRM


lib/winrm/http/auth.rb
#
# Author:: Kaustubh Deorukhkar (<kaustubh@clogeny.com>)
# Copyright:: Copyright (c) 2014 Chef Software, Inc.
# License:: Apache License, Version 2.0
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

require 'httpclient/auth'

require 'winrm/helpers/assert_patch'

def validate_patch
  # The code below patches the httpclient library to add support
  # for encrypt/decrypt as described below.
  # Add few restriction to make sure the patched methods are still
  # available, but still give a way to consciously use later versions
  PatchAssertions.assert_major_version("httpclient", [2.6,2.7], "USE_HTTPCLIENT_MAJOR")
  PatchAssertions.assert_arity_of_patched_method(HTTPClient::WWWAuth, "filter_request", 1)
  PatchAssertions.assert_arity_of_patched_method(HTTPClient::WWWAuth, "filter_response", 2)
  PatchAssertions.assert_arity_of_patched_method(HTTPClient::SSPINegotiateAuth, "set", -1)
  PatchAssertions.assert_arity_of_patched_method(HTTPClient::SSPINegotiateAuth, "get", 1)
end

# Perform the patch validations
validate_patch

# Overrides the HTTPClient::WWWAuth code to add support for encryption/decryption
# of data sent during the NTLM auth over negotiate.

# Also Overrides HTTPClient::SSPINegotiateAuth to remember user credentials, original
# library code relies on the current login user credentials on the client machine. 

# Below code helps ruby client to perform auth using the credentials provided to
# ruby client, and also enhances to use encrypted channel.

class HTTPClient

  class WWWAuth

    # Filter API implementation.  Traps HTTP request and insert
    # 'Authorization' header if needed.
    def filter_request(req)
      @authenticator.each do |auth|
        next unless auth.set? # hasn't be set, don't use it
        if cred = auth.get(req)
          auth.encrypt_payload(req) if auth.respond_to?(:encrypt_payload)
          req.header.set('Authorization', auth.scheme + " " + cred)
          return
        end
      end
    end

    # Filter API implementation.  Traps HTTP response and parses
    # 'WWW-Authenticate' header.
    #
    # This remembers the challenges for all authentication methods
    # available to the client. On the subsequent retry of the request,
    # filter_request will select the strongest method.
    def filter_response(req, res)
      command = nil
      if res.status == HTTP::Status::UNAUTHORIZED
        if challenge = parse_authentication_header(res, 'www-authenticate')
          uri = req.header.request_uri
          challenge.each do |scheme, param_str|
            @authenticator.each do |auth|
              next unless auth.set? # hasn't be set, don't use it
              if scheme.downcase == auth.scheme.downcase
                challengeable = auth.challenge(uri, param_str)
                command = :retry if challengeable
              end
            end
          end
          # ignore unknown authentication scheme
        end
      else
        decrypted_content = res.content
        @authenticator.each do |auth|
          next unless auth.set? # hasn't be set, don't use it
          decrypted_content = auth.decrypt_payload(res.content) if auth.respond_to?(:encrypted_channel?) && auth.encrypted_channel? && encrypted_content?(res.content)
        end
        # update with decrypted content
        res.content.replace(decrypted_content) if res.content and !res.content.empty?
      end
      command
    end

    private
    def encrypted_content? content
      content.start_with?('--Encrypted ')
    end
  end

  class SSPINegotiateAuth

    begin
      require 'win32/sspi'
      SSPIEnabled = true
    rescue LoadError
      SSPIEnabled = false
    end

    begin
      require 'gssapi'
      GSSAPIEnabled = true
    rescue LoadError
      GSSAPIEnabled = false
    end

    # Override to remember creds
    # Set authentication credential.
    def set(uri, user, passwd)
      # Check if user has domain specified in it.
      if user
        creds = user.split("\\")
        creds.length.eql?(2) ? (@domain,@user = creds) : @user = creds[0]
      end
      @passwd = passwd
    end

    def set?
      SSPIEnabled || GSSAPIEnabled
    end

    # Response handler: returns credential.
    # See win32/sspi for negotiation state transition.
    def get(req)
      return nil unless SSPIEnabled || GSSAPIEnabled
      target_uri = req.header.request_uri
      domain_uri, param = @challenge.find { |uri, v|
        Util.uri_part_of(target_uri, uri)
      }

      return nil unless param

      state = param[:state]
      authenticator = param[:authenticator]
      authphrase = param[:authphrase]
      case state
      when :init
        if SSPIEnabled
          # Over-ride ruby win32 sspi to support encrypt/decrypt
          require 'winrm/win32/sspi'
          authenticator = param[:authenticator] = Win32::SSPI::NegotiateAuth.new(@user, @domain, @passwd)
          @authenticator = authenticator #  **** Hacky remember as we need this for encrypt/decrypt
          return authenticator.get_initial_token
        else # use GSSAPI
          authenticator = param[:authenticator] = GSSAPI::Simple.new(domain_uri.host, 'HTTP')
          # Base64 encode the context token
          return [authenticator.init_context].pack('m').gsub(/\n/,'')
        end
      when :response
        @challenge.delete(domain_uri)
        if SSPIEnabled
          return authenticator.complete_authentication(authphrase)
        else # use GSSAPI
          return authenticator.init_context(authphrase.unpack('m').pop)
        end
      end
      nil
    end

    def encrypted_channel?
      @encrypted_channel
    end

    def encrypt_payload(req)
      if SSPIEnabled
        body = @authenticator.encrypt_payload(req.body)
        req.http_body = HTTP::Message::Body.new
        req.http_body.init_request(body)
        req.http_header.body_size = body.length if body
        # if body is encrypted update the header
        if body.include? "HTTP-SPNEGO-session-encrypted"
          @encrypted_channel = true
          req.header.set('Content-Type', "multipart/encrypted;protocol=\"application/HTTP-SPNEGO-session-encrypted\";boundary=\"Encrypted Boundary\"")
        end
      end
    end

    def decrypt_payload(body)
      body = @authenticator.decrypt_payload(body) if SSPIEnabled
      body
    end
  end
end
